File: engine/language-server/src/main/scala/org/enso/languageserver/boot/resource/SequentialResourcesInitialization.scala
package org.enso.languageserver.boot.resource
import scala.concurrent.{ExecutionContext, Future}

/** Initializes resources in sequence.
  *
  * @param resources the list of resources to initialize
  */
final class SequentialResourcesInitialization(
  resources: Seq[InitializationComponent]
)(implicit ec: ExecutionContext)
    extends InitializationComponent {

  /** @inheritdoc */
  override def isInitialized: Boolean =
    resources.forall(_.isInitialized)

  /** @inheritdoc */
  override def init(): Future[InitializationComponent.Initialized.type] =
    resources.foldLeft(Future.successful(InitializationComponent.Initialized)) {
      (action, resource) =>
        action.flatMap { _ =>
          if (resource.isInitialized)
            Future.successful(InitializationComponent.Initialized)
          else resource.init()
        }
    }
}

object SequentialResourcesInitialization {

  /** Create a sequential initialization component from a sequence of resources.
    *
    * @param resources the list of resources to initialize.
    * @param ec the execution context
    * @return the sequential initialization component
    */
  def apply(resources: InitializationComponent*)(implicit
    ec: ExecutionContext
  ): SequentialResourcesInitialization =
    new SequentialResourcesInitialization(resources)
}

File: engine/language-server/src/main/java/org/enso/languageserver/boot/resource/SequentialResourcesInitialization.java
package org.enso.languageserver.boot.resource;

import java.util.Arrays;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Executor;

/** Initializes resources in sequence. */
public class SequentialResourcesInitialization implements InitializationComponent {

  private final InitializationComponent[] resources;
  private final Executor executor;

  /**
   * Create an instance of sequential initialization component.
   *
   * @param executor the executor that runs the initialization
   * @param resources the list of resources to initialize
   */
  public SequentialResourcesInitialization(
      Executor executor, InitializationComponent... resources) {
    this.resources = resources;
    this.executor = executor;
  }

  @Override
  public boolean isInitialized() {
    return Arrays.stream(resources).allMatch(InitializationComponent::isInitialized);
  }

  @Override
  public CompletableFuture<Void> init() {
    CompletableFuture<Void> result = CompletableFuture.completedFuture(null);

    for (InitializationComponent component : resources) {
      result =
          result.thenComposeAsync(
              res ->
                  component.isInitialized()
                      ? CompletableFuture.completedFuture(null)
                      : component.init(),
              executor);
    }

    return result;
  }
}
